Connection and CASE session management over TCP

pidarped opened the pull request:
Currently, connecting to a peer is coupled with sending a message to the peer.
This decouples the two and creates a clear API for connecting to a peer address. Goes along with the existing Disconnect() API. This would be essential during activation of retained sessions by solely connecting to the peer and associating with the retained session.

Surface Connection completion and Closure callbacks and hook them through SessionManager(TransportMgr delegate) and CASESession.

Modify ActiveConnectionState in TCPBase to hold state for each connection, so that it is able to handle the various control flow paths.

Fixes https://github.com/project-chip/connectedhomeip/issues/29536
Fixes #29696 
Partially fixes #29538

andy31415 commented on src/lib/core/CHIPConfig.h:
```diff
@@ -1634,6 +1634,16 @@ extern const char CHIP_NON_PRODUCTION_MARKER[];
+ *    operational communications.
+ */
+#ifndef CHIP_CONFIG_TCP_SUPPORT_ENABLED
+#define CHIP_CONFIG_TCP_SUPPORT_ENABLED (INET_CONFIG_ENABLE_TCP_ENDPOINT)
```
I think we should only keep one variable if they are always in sync.
 
 If there is no way for TCP_SUPPORT_ENABLED and ENABLE_TCP_ENDPOINT to be different, we should drop one of them. if they can be different, a comment should explain how and why (because as a code reader I would find that surprising)

pidarped replied:
Currently only using INET_CONFIG_ENABLE_TCP_ENDPOINT everywhere.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -582,6 +589,79 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
 }
 
+#if CHIP_CONFIG_TCP_SUPPORT_ENABLED
+void SessionManager::SetConnectionCallbacks(OnConnectionCompleteCallback connCompleteCb, OnConnectionClosedCallback connClosedCb,
```
Apart from these methods, there are other snippets that are there which are TCP-specific in this file, and it would be difficult to extract all of those into a separate file.

bzbarsky-apple commented on src/app/CASESessionManager.h:
```diff
@@ -80,7 +80,7 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
      */
     void FindOrEstablishSession(const ScopedNodeId & peerId, Callback::Callback<OnDeviceConnected> * onConnection,
-                                Callback::Callback<OnDeviceConnectionFailure> * onFailure
+                                Callback::Callback<OnDeviceConnectionFailure> * onFailure, bool useTransportForLargePayload = false
```
This is a source break in public API, for consumers that use CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES, no?

pidarped replied:
Moved the new arg to the end with a default value.

bzbarsky-apple commented on src/app/CASESessionManager.cpp:
```diff
@@ -118,10 +128,11 @@ OperationalSessionSetup * CASESessionManager::FindExistingSessionSetup(const Sco
 
-Optional<SessionHandle> CASESessionManager::FindExistingSession(const ScopedNodeId & peerId) const
+Optional<SessionHandle> CASESessionManager::FindExistingSession(const ScopedNodeId & peerId,
+                                                                const bool useTransportForLargePayload) const
```
Using a bool makes the callsites pretty inscrutable....

It seems like what we really want here is an enum, probably with these values:

1. Require large payload (TCP sessions only)
2. Allow large payload (TCP or UDP; caller does not care)
3. Disallow large payload (UDP only; maybe this should be characterized in some other way?  Why would people want to ensure they get a UDP session?)

And the public API should probably reflect this set of things...

bzbarsky-apple replied:
We may have use cases that need UDP only because of latency concerns....

pidarped replied:
That is a fair point. However, we might need to think through how we need to characterize this. If vying for latency, presumably it is vanilla UDP because MRP would not serve latency needs that well. From that standpoint, what I am thinking is the following:
1. ForLargePayload(TCP sessions only)
2. LatencyOptimized(Vanilla UDP only, although there is nothing today to ensure that MRP is not adopted atop). It is up to the app to choose it, right?
3. Default(MRP or TCP) 
AllowLargePayload(TCP or UDP) does not make much sense because if it is either, it may not allow large payload over UDP, i.e.(as it stands today).
We can avoid this complexity for now and just keep it as LargePayload::Yes or No as well, and visit this later if a low latency vanilla-UDP-only-transport is desired and the layers can ensure that.

bzbarsky-apple replied:
MRP is fine from a latency perspective, in the sense that sending a new MRP message does not wait for anything previously sent.

bzbarsky-apple commented on src/app/OperationalSessionSetup.cpp:
```diff
@@ -278,6 +279,14 @@ CHIP_ERROR OperationalSessionSetup::EstablishConnection(const ReliableMessagePro
     mCASEClient = mClientPool->Allocate();
     ReturnErrorCodeIf(mCASEClient == nullptr, CHIP_ERROR_NO_MEMORY);
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
```
If this is not defined.... shouldn't the useTransportForLargePayload case just error out?

bzbarsky-apple commented on src/app/OperationalSessionSetup.h:
```diff
@@ -189,8 +189,12 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
      * cases that are detected synchronously (e.g. inability to start an address
      * lookup).
+     *
+     * `useTransportForLargePayload` is set to true when the session needs to be established
+     * over a transport that allows large payloads to be transferred, e.g., TCP.
```
Would be better to have a flags bitmap or params struct or something so the naming is clearer at the callsite.  If a callsite is passing `true`, it's hard to tell what that means without reading this code.

bzbarsky-apple commented on src/controller/CHIPDeviceController.cpp:
```diff
@@ -452,6 +452,12 @@ CHIP_ERROR DeviceCommissioner::Init(CommissionerInitParams params)
+                                                    ,
+                                                Transport::TcpListenParameters(mSystemState->TCPEndPointManager())
+                                                    .SetAddressType(IPAddressType::kIPv6)
+                                                    .SetListenPort(static_cast<uint16_t>(mUdcListenPort))
```
This makes no sense.  Why would we ever be doing UDC over TCP?

bzbarsky-apple commented on src/controller/CHIPDeviceController.h:
```diff
@@ -84,6 +84,10 @@ using namespace chip::Protocols::UserDirectedCommissioning;
 
 inline constexpr uint16_t kNumMaxActiveDevices = CHIP_CONFIG_CONTROLLER_MAX_ACTIVE_DEVICES;
 
+inline constexpr size_t kMaxUdcTransportTcpActiveConnectionCount = CHIP_CONFIG_CONTROLLER_MAX_ACTIVE_TCP_CONNECTIONS;
```
There shouldn't be any UDC over TCP.  UDC consists of messages that don't even use MRP because they absolutely do not want acks.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -638,6 +655,45 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+
+    ChipLogDetail(SecureChannel, "TCP Connection established before session establishment");
+
+    CASESession * caseSession = reinterpret_cast<CASESession *>(context);
```
What ensures that this CASESession is still alive by this point?

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -638,6 +655,45 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+    err = caseSession->SendSigma1();
+    if (err != CHIP_NO_ERROR)
+    {
+        caseSession->Clear();
```
What happens to the extra retain we did on the exchange above?  Do we just leak the exchange?  Or does the error case here end up calling Finish()?

But also: why do we need to do the extra retain on the exchange at all?  This does not make sense to me.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2131,6 +2196,15 @@ CHIP_ERROR CASESession::OnMessageReceived(ExchangeContext * ec, const PayloadHea
+            if (mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->GetPeerAddress().GetTransportType() ==
+                Transport::Type::kTcp)
+            {
+                mExchangeCtxt->Retain();
```
Why is this extra retain needed?  This does not make sense to me....

bzbarsky-apple commented on src/protocols/secure_channel/PairingSession.cpp:
```diff
@@ -60,6 +60,17 @@ void PairingSession::Finish()
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    // Release the ExchangeContext that was retained at the beginning of CASE
+    // initiation(both at initiator and responder side). The Retain() was
+    // necessary to prevent the automatic release of the ExchangeContext
+    // after sending the last message of the Exchange and before the call
```
Seems like this documentation about why the Retain() is needed should be in the callsites where Retain() is called.

That said.... clearly `mExchangeCtxt` is not usually deleted here.  Why is that different with TCP?  It shouldn't be....

bzbarsky-apple replied:
Would it be better to fix this by having our member holding the exchange be an ExchangeHandle so we never have dangling pointers as long as we don't null it out?

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -264,6 +272,11 @@ class Session
     FabricIndex mFabricIndex = kUndefinedFabricIndex;
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    // The underlying TCP connection object over which the session is
+    // established.
```
Should be clear that this will be null for sessions not over TCP, yes?  At least in API docs for the public bits.

pidarped replied:
Will enhance the comment.

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -225,6 +228,11 @@ class Session
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    ActiveConnectionState * GetTCPConnection() const { return mTCPConnection; }
+    void SetTCPConnection(ActiveConnectionState * conn) { mTCPConnection = conn; }
```
Having this Set() be callable any time is a bit odd, but I guess sessions can migrate to new connections?

What is the lifetime situation here?  How do we avoid dangling pointers to ActiveConnectionState?

bzbarsky-apple replied:
The lifetime assumptions need to be clearly documented.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,106 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+        ChipLogProgress(Inet, "Connection Closed callback missing");
+    }
+
+    // Mark the corresponding secure session as defunct
```
```suggestion
    // Mark the corresponding secure sessiona as defunct
```

bzbarsky-apple commented on src/transport/SessionManager.h:
```diff
@@ -477,6 +500,12 @@ class DLL_EXPORT SessionManager : public TransportMgrDelegate, public FabricTabl
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    OnConnectionCompleteCallback mConnCompleteCb = nullptr;
+    OnConnectionClosedCallback mConnClosedCb     = nullptr;
+    void * mConnContext                          = nullptr;
```
This makes no sense, in that CASESession sets this when sending Sigma1.  There's one SessionManager, but there can be (and are, for real clients like hubs) many parallel CASE establishments happening at once.

Put another way: when a hub boots, it will in parallel try to do CASE with all the devices.  That will totally break here.

pidarped replied:
This was an oversight. The SessionManager needs to handle multiple sessions. One method would be to pass an appContext down and get it back on connection completion to be used to multiplex among the various sessions at the SessionManager.

pidarped replied:
Created an App context that is passed in with ConnectToPeer() and plumbed all the way to the Connection context object and returned on connection completion to get to the right CASESession.

bzbarsky-apple commented on src/transport/TransportMgr.h:
```diff
@@ -49,8 +52,24 @@ class TransportMgrDelegate
      */
-    virtual void OnMessageReceived(const Transport::PeerAddress & source, System::PacketBufferHandle && msgBuf) = 0;
+    virtual void OnMessageReceived(const Transport::PeerAddress & source, System::PacketBufferHandle && msgBuf,
+                                   Transport::MessageTransportContext * ctxt = nullptr) = 0;
```
This is a change to a public API, right?  Or is this not public API?

pidarped replied:
I believe it is. This is necessary to know which connection the message arrived on. Otherwise, the implicit assumption would be that there is exactly one connection to a given peer.

bzbarsky-apple replied:
My point is that we might want to think about making this change in a compatible way: have both the new and old signatures, so old API clients which presumably don't care about the context don't get it, and new ones know to override the new signature, not the old one.

And default impl of the new signature calls the old signature.

pidarped replied:
Agreed. I was resorting to the default arg value of nullptr to the connection context to essentially achieve that. For old api clients, the argument should be not present in the calls and default to null always.

bzbarsky-apple replied:
> For old api clients, the argument should be not present in the calls and default to null always

Old API clients _implement this method_, no?  So this is a compile-time break for them as currently done.

pidarped replied:
Ok. Apologies for the miscommunication. This is not a SDK public API. It is a public API of the SessionManager and this change terminates in the SessionManager itself. So, beyond that, when the message is sent up to the ExchangeMgr, this is not carried.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,106 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+            session->SetTCPConnection(nullptr);
+            // Mark session as defunct
+            session->MarkAsDefunct();
+            return Loop::Break;
```
There should ideally be one active session over a given connection.

bzbarsky-apple replied:
"Ideally", or guaranteed?

pidarped replied:
I would think it to be guaranteed. However, the number of sessions should be relatively small and it may not harm in continuing through the loop, just in case.

pidarped commented on src/transport/SessionManager.h:
```diff
@@ -52,8 +52,22 @@
+ * transport that supports large messages.
+ *
+ */
+enum class UseLargePayloadTransport : uint8_t
```
Your point is valid on the extensibility of this enum. Renamed it to TransportPayloadCapability

yufengwangca commented on src/app/CASESessionManager.cpp:
```diff
@@ -146,7 +153,16 @@ void CASESessionManager::ReleaseAllSessions()
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CASESessionManager::GetPeerAddressForLargePayloadSession(const ScopedNodeId & peerId, Transport::PeerAddress & addr)
```
nit: GetPeerAddressFromLargePayloadSession ?

pidarped replied:
I think GetPeerAddressForLargePayloadSession() seems fine. I wanted it to mean to get the peer address for a session that supports large payload for the given PeerId.

yufengwangca commented on src/app/OperationalSessionSetup.h:
```diff
@@ -202,8 +202,12 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
      * cases that are detected synchronously (e.g. inability to start an address
      * lookup).
+     *
+     * `transportPayloadCapability` is set to true when the session needs to be established
```
is set to kLarge

yufengwangca commented on src/app/OperationalSessionSetup.h:
```diff
@@ -219,8 +223,12 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
      * `onSetupFailure` may be called before the Connect call returns, for error cases that are detected synchronously
      * (e.g. inability to start an address lookup).
+     *
+     * `transportPayloadCapability` is set to true when the session needs to be established
```
is set to kLarge

yufengwangca commented on src/app/OperationalSessionSetup.h:
```diff
@@ -305,6 +313,8 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
 
     bool mPerformingAddressUpdate = false;
 
+    TransportPayloadCapability mSessionForLargePayload = TransportPayloadCapability::kDefault;
```
mSessionForLargePayload does not reflect the meaning of TransportPayloadCapability well, maybe mTransportPayloadCapability or  mIsSessionForLargePayload

yufengwangca commented on src/app/OperationalSessionSetup.cpp:
```diff
@@ -291,6 +295,21 @@ CHIP_ERROR OperationalSessionSetup::EstablishConnection(const ReliableMessagePro
     mCASEClient = mClientPool->Allocate();
     ReturnErrorCodeIf(mCASEClient == nullptr, CHIP_ERROR_NO_MEMORY);
 
+    // TODO: Combine LargePayload flag with DNS-SD advertisements from peer
```
Create a ticket to track this task

pidarped replied:
https://github.com/project-chip/connectedhomeip/issues/29539 should be usable for this.

yufengwangca commented on src/app/OperationalSessionSetup.cpp:
```diff
@@ -291,6 +295,21 @@ CHIP_ERROR OperationalSessionSetup::EstablishConnection(const ReliableMessagePro
+        // Set the transport type for carrying large payloads
+        mDeviceAddress.SetTransportType(chip::Transport::Type::kTcp);
+#else
+        ChipLogError(Discovery, "Cannot setup session for large payload. Appropriate Transport unavailable");
```
What about kBle, do we use connection oriented mode of BLE for Matter

pidarped replied:
For operational communication BLE is not used. It is only used for Commissioning.

yufengwangca commented on src/controller/CHIPDeviceControllerSystemState.h:
```diff
@@ -57,16 +57,25 @@ namespace chip {
-                                        ,
-                                        Transport::UDP /* IPv4 */
+                 ,
+                 Transport::UDP /* IPv4 */
```
Why specify 'Transport::UDP' twice for IPv4 and IPv6? Does TCP only supports IPv6?

pidarped replied:
The UDP for IPv4 is legacy and I kept it as-is. For TCP, I have only defined it for IPv6. We need to look into if the UDP over IPv4 is required or not.

yufengwangca commented on src/messaging/ExchangeContext.cpp:
```diff
@@ -670,5 +670,12 @@ void ExchangeContext::ExchangeSessionHolder::GrabExpiredSession(const SessionHan
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void ExchangeContext::OnSessionConnectionClosed(CHIP_ERROR conErr)
+{
+    // TODO: Handle connection closure at the ExchangeContext level.
```
Need to create a ticket to track the pending task

pidarped replied:
Filed https://github.com/project-chip/connectedhomeip/issues/32250.

yufengwangca commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -645,6 +669,74 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
+    VerifyOrReturn(caseSession != nullptr);
+
+    // Exit and disconnect if connection setup encountered an error.
+    SuccessOrExit(err);
```
move the parameter check to the beginning of the function.

pidarped replied:
It is almost at the beginning. The CASESession object is just obtained before this so that it can be used in the exit label. I will add a log message there with the error.

tcarmelveilleux commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,146 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    }
+    else
+    {
+        // TODO: Refactor to ensure that the Disconnection is for the
```
Add issue # for each TODO

tcarmelveilleux commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,146 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    {
+        // TODO: Refactor to ensure that the Disconnection is for the
+        // corresponding higher layer entity that tried to set up the connection
+        ChipLogProgress(Inet, "Connection Complete callback missing");
```
We should remove the "callback" logging, and only log "connection established", for example. We should also log the IP address and port.

pidarped replied:
I have removed the logging for the path where the callback is called(success path), but kept the logging with added IP address/port information since the connection is going to be disconnected in this case.

tcarmelveilleux replied:
I meant that talking about the "callback" itself is much less relevant than logging the edges of important events.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,146 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
     else
     {
-        UnauthenticatedMessageDispatch(partialPacketHeader, peerAddress, std::move(msg));
+        ChipLogProgress(Inet, "Connection Closed callback missing");
```
Removed this log.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,146 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+{
+    if (mTransportMgr != nullptr)
+    {
+        ChipLogProgress(Inet, "Disconnecting from peer ");
```
Done.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,146 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+        return mTransportMgr->ConnectToPeer(peerAddress, appState);
+    }
+
+    ChipLogError(Inet, "The transport manager is not initialized. Unable to connect to peer");
```
Added.

tcarmelveilleux commented on src/app/CASESessionManager.h:
```diff
@@ -83,7 +84,8 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
 #endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
-    );
+                                ,
+                                TransportPayloadCapability transportPayloadCapability = TransportPayloadCapability::kDefault);
```
What is "default". It's  a very relative name. Is there a better name?

pidarped replied:
Default was meant to indicate the MRP path which is, essentially, the default path with a fixed MTU defined payload size limit. I could change it to MRPPayload if that helps clarify it better.

pidarped replied:
Changed the kDefault to kMRPPayload.

bzbarsky-apple commented on src/app/CASESessionManager.h:
```diff
@@ -106,14 +108,16 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
      * @param onSetupFailure A callback to be called upon an extended device connection failure.
      * @param attemptCount The number of retry attempts if session setup fails (default is 1).
      * @param onRetry A callback to be called on a retry attempt (enabled by a config flag).
+     * @param transportPayloadCapability An indicator of whether to use a transport that supports transfer of large payloads.
```
```suggestion
     * @param transportPayloadCapability An indicator of what sorts of payloads the session needs to be able to transport.
```

bzbarsky-apple commented on src/app/CASESessionManager.h:
```diff
@@ -134,13 +138,15 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
      * @param onConnection A callback to be called upon successful connection establishment.
      * @param attemptCount The number of retry attempts if session setup fails (default is 1).
      * @param onRetry A callback to be called on a retry attempt (enabled by a config flag).
+     * @param transportPayloadCapability An indicator of whether to use a transport that supports transfer of large payloads.
```
See above.

bzbarsky-apple commented on src/app/CASESessionManager.h:
```diff
@@ -156,6 +162,8 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
      */
     CHIP_ERROR GetPeerAddress(const ScopedNodeId & peerId, Transport::PeerAddress & addr);
 
+    CHIP_ERROR GetPeerAddressForLargePayloadSession(const ScopedNodeId & peerId, Transport::PeerAddress & addr);
```
Why wouldn't we just have a TransportPayloadCapability argument to GetPeerAddress, either via defaulted arg or via overload?

pidarped replied:
Added as a defaulted arg to GetPeerAddress().

bzbarsky-apple commented on src/app/CASESessionManager.h:
```diff
@@ -165,15 +173,18 @@ class CASESessionManager : public OperationalSessionReleaseDelegate, public Sess
                                       uint8_t attemptCount, Callback::Callback<OnDeviceConnectionRetry> * onRetry
 #endif
-    );
+                                      ,
```
This looks like it will not compile when CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES is false, to me.

bzbarsky-apple commented on src/app/OperationalSessionSetup.cpp:
```diff
@@ -291,6 +295,22 @@ CHIP_ERROR OperationalSessionSetup::EstablishConnection(const ReliableMessagePro
+        // Set the transport type for carrying large payloads
+        mDeviceAddress.SetTransportType(chip::Transport::Type::kTcp);
+#else
+        ChipLogError(Discovery, "Cannot setup session for large payload. Appropriate Transport unavailable");
```
```suggestion
        ChipLogError(Discovery, "Cannot set up session for large payload. Appropriate Transport unavailable");
```

bzbarsky-apple commented on src/app/OperationalSessionSetup.cpp:
```diff
@@ -291,6 +295,22 @@ CHIP_ERROR OperationalSessionSetup::EstablishConnection(const ReliableMessagePro
+#else
+        ChipLogError(Discovery, "Cannot setup session for large payload. Appropriate Transport unavailable");
+
+        CleanupCASEClient();
```
Why aren't we just doing this check before allocating mCASEClient?

pidarped replied:
Moved to before allocation.

bzbarsky-apple commented on src/app/OperationalSessionSetup.h:
```diff
@@ -202,8 +202,12 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
      * cases that are detected synchronously (e.g. inability to start an address
      * lookup).
+     *
+     * `transportPayloadCapability` is set to kLarge when the session needs to be established
```
There is no kLarge value.

bzbarsky-apple commented on src/app/OperationalSessionSetup.h:
```diff
@@ -219,8 +223,12 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
      * `onSetupFailure` may be called before the Connect call returns, for error cases that are detected synchronously
      * (e.g. inability to start an address lookup).
+     *
+     * `transportPayloadCapability` is set to kLarge when the session needs to be established
```
No kLarge.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -646,6 +670,79 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+
+    Transport::PeerAddress peerAddr = conn->mPeerAddr;
+    peerAddr.ToString(peerAddrBuf);
+    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
```
What guarantees that conn->mAppState is still alive at this point?  These sorts of things need to be clearly documented.

pidarped replied:
SessionManager checks for it. Added a VerifyOrDie for it to be non-null anyhow.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -646,6 +670,79 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+    {
+        ChipLogError(SecureChannel, "Connection establishment failed: %s", ErrorStr(err));
+
+        // Release the retained ExchangeContext above if an error happened.
+        caseSession->mExchangeCtxt->Release();
```
We can get here with err an error _without having made the retain call_.  

Can we please write this code in a less error-prone way?  What happened with my suggestion that we use an ExchangeHandle to hold the exchange, so we are holding a reference to it, period?

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -646,6 +670,79 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+{
+    VerifyOrReturn(conn != nullptr);
+
+    CASESession * caseSession = reinterpret_cast<CASESession *>(conn->mAppState->appContext);
```
How do we know the objects are still alive?

pidarped replied:
Added a VerifyOrDie() for the conn->mAppState to be non-null. SessionManager should already have a check.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -646,6 +670,79 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
 }
 
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+void CASESession::HandleConnectionComplete(Transport::ActiveConnectionState * conn, CHIP_ERROR err)
```
What happens if connection setup fails?  Is this the method that gets called?  If so, it's not named very well....  Should be HandleConnectionAttemptComplete or something.

pidarped replied:
Yes, this method gets called on setup failure. Renamed it.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2224,6 +2336,18 @@ CHIP_ERROR CASESession::OnMessageReceived(ExchangeContext * ec, const PayloadHea
+        if (mExchangeCtxt->GetSessionHandle()->AsUnauthenticatedSession()->GetPeerAddress().GetTransportType() ==
+            Transport::Type::kTcp)
+        {
+            mExchangeCtxt->Release();
```
Again, we can be here without having done the retain.

pidarped replied:
Changed to ExchangeHandle.

bzbarsky-apple commented on src/protocols/secure_channel/PairingSession.cpp:
```diff
@@ -58,6 +59,24 @@ void PairingSession::Finish()
+        // Associate the connection with the secure session being activated.
+        mSecureSessionHolder->AsSecureSession()->SetTCPConnection(conn);
+
+        mExchangeCtxt->Release();
```
What guarantees that we reached one of the Retain() calls before now?  I suspect nothing.

pidarped replied:
Changed to ExchangeHandle.

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -225,6 +228,19 @@ class Session
+    // latter is about to be marked active. It is also used to reset the
+    // connection to a nullptr when the connection is lost and the session
+    // is marked as Defunct.
+    // The lifetime of this member connection pointer is, essentially the same
```
```suggestion
    // The lifetime of this member connection pointer is, essentially, the same
```

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -225,6 +228,19 @@ class Session
+    // is marked as Defunct.
+    // The lifetime of this member connection pointer is, essentially the same
+    // as that of the underlying connection with the peer.
+    // It would remain as a nullptr for all sessions that are not set up over
+    // a TCP connection.
```
This part, and probably the lifetime part, belong with the getter, no?

pidarped replied:
Consolidated the Getter and Setter together.

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -264,6 +280,15 @@ class Session
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    // The underlying TCP connection object over which the session is
+    // established.
+    // The lifetime of this member connection pointer is, essentially the same
```
Why are we duplicating these comments?

pidarped replied:
Fixed.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+
+    VerifyOrReturn(conn != nullptr);
+
+    appConnCbCtxt = conn->mAppState;
```
Why is this not just declared here.  Please declare things where used, not at the tops of methods.

pidarped replied:
Done.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    else
+    {
+        ChipLogProgress(Inet, "%s: Received connection request from %s, but no registered handler. Disconnecting.", peerAddrBuf,
+                        __func__);
```
Please don't use `__func__` in possibly-embedded code.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+
+void SessionManager::HandleConnectionComplete(Transport::ActiveConnectionState * conn, CHIP_ERROR conErr)
+{
+    Transport::AppConnCallbackCtxt * appConnCbCtxt = nullptr;
+    char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
```
Again, declare where used.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    else
+    {
+        ChipLogProgress(Inet, "%s: Connection established with peer %s, but no registered handler. Disconnecting.", peerAddrBuf,
+                        __func__);
```
No `__func__`

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+
+void SessionManager::HandleConnectionClosed(Transport::ActiveConnectionState * conn, CHIP_ERROR conErr)
+{
+    Transport::AppConnCallbackCtxt * appConnCbCtxt = nullptr;
```
Declare where used.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    {
+        char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
+        peerAddress.ToString(peerAddrBuf);
+        ChipLogProgress(Inet, "Connecting to peer at %s.", peerAddrBuf);
```
Should probably mention TCP in there somewhere.

pidarped replied:
Added TCP in the log. The addresstype for PeerAddress emits the transport protocol in the log as well.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+        return mTransportMgr->ConnectToPeer(peerAddress, appState);
+    }
+
+    ChipLogError(Inet, "The transport manager is not initialized. Unable to connect to peer.");
```
Should probably log which peer here too.

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -573,20 +579,162 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    {
+        char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
+        peerAddress.ToString(peerAddrBuf);
+        ChipLogProgress(Inet, "Disconnecting from peer at %s.", peerAddrBuf);
```
These various logs here and in the Disconnect below should perhaps mention TCP.

pidarped replied:
Added in the log. The Peeraddress also prints TCP as the transport type. We can revisit later as to whether the extra explicit mention is needed.

bzbarsky-apple commented on src/transport/SessionManager.h:
```diff
@@ -52,8 +52,26 @@
+ */
+enum class TransportPayloadCapability : uint8_t
+{
+    kMRPPayload,  // This payload limits the size to be bound within an IPv6
+                  // packet size of 1280 bytes.
```
I don't know what "bound within" means.

Bu also: this is not addressing the latency issue I raised months ago.  Consumers will want to be able to express "TCP", "MRP" and "Either one", and this is not doing that.

pidarped replied:
I think I responded to your earlier comment about this. So, if the user wants to choose a low-latency transport, they should choose plain UDP. If it is plain UDP, then 'Either One' may not capture it correctly and there is additional plumbing underneath needed for that, presumably. This was the reason I left it as-is for now. If the goal of the third option is to use any transport that has a session up, that we could do, even if MRPPayload was chosen since this enum is about specifying the size and any transport that can handle it should suffice.

bzbarsky-apple commented on src/transport/SessionManager.h:
```diff
@@ -52,8 +52,26 @@
+{
+    kMRPPayload,  // This payload limits the size to be bound within an IPv6
+                  // packet size of 1280 bytes.
+    kLargePayload // This payload indicates a large size beyond the size limit
+                  // restrictions of an IPv6 packet as imposed by the MRP
+                  // transport. The transport of choice, in this case, is TCP.
```
```suggestion
    kLargePayload // The transport needs to handle payloads larger
                  // than the single IPv6 packet supported by an MRP
                  // transport. The transport of choice, in this case, is TCP.
```

bzbarsky-apple commented on src/transport/raw/ActiveConnectionState.h:
Should this be called ActiveTCPConnectionState.h?

pidarped replied:
Renamed.

bzbarsky-apple commented on src/transport/raw/ActiveConnectionState.h:
```diff
@@ -0,0 +1,119 @@
+namespace Transport {
+
+/**
+ *  The State of the TCP connection
```
This thing is called Transport::State.  There is no mention of TCP in the naming.  There probably needs to be.

pidarped replied:
Renamed to TCPState.

bzbarsky-apple commented on src/transport/raw/ActiveConnectionState.h:
```diff
@@ -0,0 +1,119 @@
+    kClosed      = 5, /**< Connection is closed. */
+};
+
+struct AppConnCallbackCtxt;
```
This is a pretty cryptic name, and again just putting in in Transport:: is strange.

pidarped replied:
Changed to AppTCPConnectionCallbackCtxt

bzbarsky-apple commented on src/transport/raw/ActiveConnectionState.h:
```diff
@@ -0,0 +1,119 @@
+/**
+ *  State for each active connection
+ */
+struct ActiveConnectionState
```
Naming should probably mention TCP, if it's TCP-specific.

bzbarsky-apple commented on src/transport/raw/ActiveConnectionState.h:
```diff
@@ -0,0 +1,119 @@
+};
+
+// Functors for callbacks into higher layers
+using OnConnectionReceivedCallback = void (*)(ActiveConnectionState * conn);
+
+using OnConnectionCompleteCallback = void (*)(ActiveConnectionState * conn, CHIP_ERROR conErr);
+
+using OnConnectionClosedCallback = void (*)(ActiveConnectionState * conn, CHIP_ERROR conErr);
```
Again, if these are all TCP-specific they should be named accordingly.

pidarped replied:
Renamed with TCP in the name.

pidarped replied:
Added TCP in the names.

bzbarsky-apple commented on src/transport/raw/Base.h:
```diff
@@ -77,11 +95,26 @@ class Base
+    /**
+     * Connect to the specified peer.
+     */
+    virtual CHIP_ERROR ConnectToPeer(const PeerAddress & address, Transport::AppConnCallbackCtxt * appState)
```
Should mention TCP in naming, if it's TCP-specific.

pidarped replied:
There was a Disconnect() API for TCP that was pre-existing, and this was aligned with that. TCP was expected to be implied here. Do you prefer it be TCPConnectToPeer() and TCPDisconnect() or something like that?

pidarped replied:
Renamed the APIs around connecting/disconnecting to TCPConnect()/TCPDisconnect().

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
     bool foundPendingPacket = false;
     TCPBase * tcp           = reinterpret_cast<TCPBase *>(endPoint->mAppState);
     Inet::IPAddress ipAddress;
+    char ipAddrStr[64];
```
Where is this 64 coming from?  Looks suspect to me.  Shouldn't this be the relevant constant from IPAddress?

pidarped replied:
Changed to IPAddress constant.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
     endPoint->GetPeerInfo(&ipAddress, &port);
     endPoint->GetInterfaceId(&interfaceId);
     PeerAddress addr = PeerAddress::TCP(ipAddress, port, interfaceId);
+    ipAddress.ToString(ipAddrStr, sizeof(ipAddrStr));
```
Does that method not have an overload taking a char array?

pidarped replied:
Yes, it does. Initially I thought it didn't. Removed the sizeof.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
-            endPoint->Free();
-            ChipLogError(Inet, "Internal logic error: insufficient space to store active connection");
-        }
+        ChipLogError(Inet, "Connection establishment with [%s:%d] encountered an error: %s", ipAddrStr, port, ErrorStr(err));
```
Please use CHIP_ERROR_FORMAT instead of forcing strings.

pidarped replied:
Done.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+    ActiveConnectionState * activeConnection = tcp->FindActiveConnection(endPoint);
+    Inet::IPAddress ipAddress;
+    uint16_t port;
+    char ipAddrStr[64];
```
Again, where is the 64 coming from?

pidarped replied:
Changed to IPAddress specific const.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+    char ipAddrStr[64];
+
+    endPoint->GetPeerInfo(&ipAddress, &port);
+    ipAddress.ToString(ipAddrStr, sizeof(ipAddrStr));
```
See above.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -359,126 +422,171 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+    Inet::InterfaceId interfaceId;
+    Inet::IPAddress ipAddress;
+    uint16_t port;
+    char ipAddrStr[64];
```
See comments above, and for ToString down below.

Also, why is this not declared where it's used?

pidarped replied:
Moved the declaration to where it is used.

bzbarsky-apple commented on src/transport/raw/TCPConfig.h:
```diff
@@ -0,0 +1,144 @@
+/**
+ * @def CHIP_CONFIG_DEVICE_MAX_ACTIVE_TCP_CONNECTIONS
+ *
+ * @brief Maximum Number of TCP connections a device can simultaneously have
```
Why is this a separate ifdef from "controller"?

What are the definitions of "device" and "controller" here for a single node that would use both ifdefs?  I have no idea how I would set these values for a node I am creating.

pidarped replied:
Collapsed into one ifdef which should be settable by both device and controller at their respective ends.

bzbarsky-apple commented on src/transport/raw/tests/TestTCP.cpp:
I did not review this file.

bzbarsky-apple reviewed (changes requested):
The mismatched retains are not OK.

pidarped commented on src/app/CASESessionManager.cpp:
```diff
@@ -35,14 +35,16 @@ void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Cal
                                                 uint8_t attemptCount, Callback::Callback<OnDeviceConnectionRetry> * onRetry
 #endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
-)
+                                                ,
```
Done.

pidarped replied:
Done.

pidarped committed:
TCP connection setup/management and CASESession association.

Add TCPConnect()/TCPDisconnect() API for explicit connection setup.
Currently, connecting to a peer is coupled with sending a message to the
peer.
This decouples the two and creates a clear API for connecting to a peer
address. Goes along with the existing Disconnect() API.
This would be essential during activation of retained sessions by solely
connecting to the peer and associating with the retained session.

Surface Connection completion and Closure callbacks and hook them
through SessionManager(TransportMgr delegate) and CASESession.

Mark SecureSession as defunct on connection closures.

Modify ActiveConnectionState in TCPBase to hold state for each
connection, so that it is able to handle the various control flow paths.

Associate a session with a connection object.

Associate the PeerAddress with the session early. Pass the PeerAddress
in the Find APIs. This helps check against the correct TransportType
when searching for a Sesssion in the SessionTable.

Add a `large payload` flag in EstablishSession() and Session lookup
functions to create/associate with the correct session and transport.

Have default configurations for TCP in a separate TCPConfig.h.

Refactor echo_requester.cpp and echo_responder.cpp to use the session
associated with the connection.

Handle Connection closure at ExchangeMgr and uplevel to corresponding
ExchangeContext using the corresponding session handle.

Add tests around connection establishment in TestTCP.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -353,133 +422,180 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+
+        char ipAddrStr[Inet::IPAddress::kMaxStringLength];
+        peerAddress.ToString(ipAddrStr);
+        ChipLogProgress(Inet, "Incoming connection established with peer at [%s:%d]", ipAddrStr, port);
```
Why is this not logging peerAddress?

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -353,133 +422,180 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
-        endPoint->OnConnectionReceived = OnConnectionReceived;
-        endPoint->OnAcceptError        = OnAcceptError;
-        endPoint->OnPeerClose          = OnPeerClosed;
+        // By default, disable TCP Nagle buffering by setting TCP_NODELAY socket option to true
```
This should be documenting _why_ it's doing that.

bzbarsky-apple commented on src/transport/raw/TCP.h:
```diff
@@ -153,21 +133,38 @@ class DLL_EXPORT TCPBase : public Base
     CHIP_ERROR SendMessage(const PeerAddress & address, System::PacketBufferHandle && msgBuf) override;
 
-    void Disconnect(const PeerAddress & address) override;
+    CHIP_ERROR TCPConnect(const PeerAddress & address, Transport::AppTCPConnectionCallbackCtxt * appState,
+                          Transport::ActiveTCPConnectionState ** peerConnState) override;
```
This needs some documentation about what the outparam means (and that it's an outparam), and how it can be used safely.

bzbarsky-apple commented on src/transport/raw/TCP.h:
```diff
@@ -153,21 +133,38 @@ class DLL_EXPORT TCPBase : public Base
+    // Close an active connection (corresponding to the passed
+    // ActiveTCPConnectionState object)
+    // and release from the pool.
+    void TCPDisconnect(Transport::ActiveTCPConnectionState * conn, bool shouldAbort = 0) override;
```
```suggestion
    void TCPDisconnect(Transport::ActiveTCPConnectionState * conn, bool shouldAbort = false) override;
```

pidarped commented on src/transport/Session.h:
```diff
@@ -264,6 +276,15 @@ class Session
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+    // The underlying TCP connection object over which the session is
+    // established.
+    // The lifetime of this member connection pointer is, essentially the same
```
Resolved now. I thought I did it last time, but can't be sure.

pidarped commented on src/transport/SessionConnectionDelegate.h:
```diff
@@ -0,0 +1,47 @@
+ * @brief
+ *   This class provides a skeleton for the callback functions related to sessions over TCP,
+ *   when the underlying connection changes. The functions will be called by SessionManager
+ *   object on specific events. If the user of SessionManager is interested in receiving
```
Done.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -587,20 +593,155 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
+
+    VerifyOrReturn(conn != nullptr);
+    Transport::PeerAddress peerAddr = conn->mPeerAddr;
```
Removed temporary.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -587,20 +593,155 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    }
+    else
+    {
+        ChipLogProgress(Inet, "TCP Connection established with peer %s, but no registered handler. Disconnecting.", peerAddrBuf);
```
Moved to inside block.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -587,20 +593,155 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
+    if (mTransportMgr != nullptr)
+    {
+        char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
+        Transport::PeerAddress peerAddr = conn->mPeerAddr;
```
Removed temporary.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -689,13 +850,50 @@ void SessionManager::UnauthenticatedMessageDispatch(const PacketHeader & partial
+        if (sessionConn == nullptr)
+        {
+            secureSession->SetTCPConnection(ctxt->conn);
+            secureSession->SetPeerAddress(peerAddress);
```
Done.

pidarped commented on src/transport/SessionManager.cpp:
```diff
@@ -1021,13 +1211,19 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & partialPack
-            (!type.HasValue() || type.Value() == session->GetSecureSessionType()))
+            (!type.HasValue() || type.Value() == session->GetSecureSessionType())
+#if INET_CONFIG_ENABLE_TCP_ENDPOINT
+            && (transportPayloadCapability == TransportPayloadCapability::kLargePayload ? (session->GetTCPConnection() != nullptr)
```
Handled the 3rd option.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.h:
```diff
@@ -286,6 +286,22 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
+    // Context to pass down when connecting to peer
+    Transport::AppTCPConnectionCallbackCtxt mTCPConnCbCtxt;
+    // Pointer to the underlying TCP connection state. Returned by the
+    // TCPConnect() method(on the connection Initiator side) when an
```
```suggestion
    // TCPConnect() method (on the connection Initiator side) when an
```

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.h:
```diff
@@ -286,6 +286,22 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
+    Transport::AppTCPConnectionCallbackCtxt mTCPConnCbCtxt;
+    // Pointer to the underlying TCP connection state. Returned by the
+    // TCPConnect() method(on the connection Initiator side) when an
+    // ActiveTCPConenctionState object is allocated. This may also be assigned
```
```suggestion
    // ActiveTCPConnectionState object is allocated. This may also be assigned
```

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.h:
```diff
@@ -286,6 +286,22 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
+    // TCPConnect() method(on the connection Initiator side) when an
+    // ActiveTCPConenctionState object is allocated. This may also be assigned
+    // to the connection allocated on the receiver side when an incoming
+    // connection request arrives, and invalidated when the connection is
+    // closed.
```
```suggestion
    // connection request arrives.
    //
    // This pointer must be nulled out when the connection is closed.
```

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -648,6 +687,83 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+    }
+}
+
+void CASESession::HandleConnectionClosed(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR conErr)
```
Will this get called if we ourselves call TCPDisconnect()?  This is very non-obvious to me so far.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2140,10 +2256,16 @@ CHIP_ERROR CASESession::OnMessageReceived(ExchangeContext * ec, const PayloadHea
+        mExchangeCtxt.Value()->GetSessionHandle()->AsUnauthenticatedSession()->GetPeerAddress().GetTransportType() !=
+            Transport::Type::kTcp)
+    {
+        // TODO: Rename FlushAcks() to something more semantically correct and
```
Please make sure a followup issue is filed to track fixing up this stuf and have this code comment point to that issue.

pidarped replied:
Filed issue #33183

bzbarsky-apple commented on src/transport/SessionManager.cpp:
```diff
@@ -1057,27 +1240,59 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & partialPack
+
+    if (transportPayloadCapability == TransportPayloadCapability::kMRPOrTCPCompatiblePayload)
+    {
+        return mrpSession != nullptr
+            ? MakeOptional<SessionHandle>(*mrpSession)
+            : (tcpSession != nullptr ? MakeOptional<SessionHandle>(*tcpSession) : Optional<SessionHandle>::Missing());
```
This is really hard to follow.  Maybe just write it out with `if` and multiple return statements?

pidarped replied:
Separated into multiple ifs and returns.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -353,171 +439,243 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+    }
 
-    tcp->ReleaseActiveConnection(endPoint);
+    if (conn->IsConnected())
```
I am not quite following this part.  Doesn't CASESession call TCPDisconnect on a connection that might still be getting set up?

pidarped replied:
Yes, it does. Refactored to include a check for IsConnecting() as well.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -648,6 +687,83 @@ CHIP_ERROR CASESession::RecoverInitiatorIpk()
+    {
+        ChipLogError(SecureChannel, "Connection establishment failed: %s", ErrorStr(err));
+
+        // Close the underlying connection
```
Ah, excellent.  Thank you.

bzbarsky-apple commented on src/transport/raw/TCP.cpp:
```diff
@@ -353,171 +439,243 @@ CHIP_ERROR TCPBase::OnTcpReceive(Inet::TCPEndPoint * endPoint, System::PacketBuf
+            err = mActiveConnections[i].mEndPoint->GetPeerInfo(&ipAddress, &port);
+            if (err != CHIP_NO_ERROR)
+            {
+                ChipLogError(Inet, "GetPeerInfo error: %" CHIP_ERROR_FORMAT, err.Format());
```
Followup: These error logs in this function should make it clearer that this is while trying to disconnect TCP.

pidarped replied:
Prepended with TCPDisconnect: to make it clear.
